File: packages/shared-python/security/src/security/documents.py
# validates the identity documents a ticket holder or an account can be named with
import enum
import re

_DNI_RE = re.compile(r"^\d{8}[A-Z]$")
_NIE_RE = re.compile(r"^[XYZ]\d{7}[A-Z]$")
_OTHER_RE = re.compile(r"^[A-Z0-9]{5,20}$")
_LETTER_MAP = "TRWAGMYFPDXBNJZSQVHLCKE"
_NIE_PREFIX = {"X": "0", "Y": "1", "Z": "2"}


class DocumentType(enum.StrEnum):
    dni = "dni"
    nie = "nie"
    other = "other"


# checks the control letter a spanish identity number ends with
def _valid_control_letter(digits: str, letter: str) -> bool:
    return _LETTER_MAP[int(digits) % 23] == letter


# validates a document against the rules of the type it claims to be
def validate_document(value: str, document_type: DocumentType | str) -> str:
    v = value.strip().upper().replace(" ", "").replace("-", "")
    kind = DocumentType(document_type)

    if kind is DocumentType.dni:
        if not _DNI_RE.match(v):
            raise ValueError("A DNI is 8 digits followed by a letter.")
        if not _valid_control_letter(v[:8], v[8]):
            raise ValueError("DNI check letter rejected.")
        return v

    if kind is DocumentType.nie:
        if not _NIE_RE.match(v):
            raise ValueError("A NIE is X, Y or Z followed by 7 digits and a letter.")
        if not _valid_control_letter(_NIE_PREFIX[v[0]] + v[1:8], v[8]):
            raise ValueError("NIE check letter rejected.")
        return v

    # a foreign document has no check digit that holds across every issuing country,
    # so only its shape is checked here and a person reviews the photo behind it
    if not _OTHER_RE.match(v):
        raise ValueError("A document number is 5 to 20 letters or digits.")
    return v


# reports the type a spanish document belongs to when no type was captured with it
def infer_document_type(value: str) -> DocumentType:
    v = value.strip().upper()
    if _DNI_RE.match(v):
        return DocumentType.dni
    if _NIE_RE.match(v):
        return DocumentType.nie
    return DocumentType.other

File: apps/app/src/lib/documents.ts
// mirrors the document rules the backend enforces so the form can warn as you type
export type DocumentType = 'dni' | 'nie' | 'passport'

export const DOCUMENT_TYPES: DocumentType[] = ['dni', 'nie', 'passport']

const LETTERS = 'TRWAGMYFPDXBNJZSQVHLCKE'
const DNI_RE = /^\d{8}[A-Z]$/
const NIE_RE = /^[XYZ]\d{7}[A-Z]$/
const PASSPORT_RE = /^[A-Z0-9]{6,12}$/
const NIE_PREFIX: Record<string, string> = { X: '0', Y: '1', Z: '2' }

// normalises a typed document the same way the backend does before checking it
export function normaliseDocument(value: string): string {
  return value.trim().toUpperCase().replace(/[\s-]/g, '')
}

// checks the control letter a spanish identity number ends with
function validControlLetter(digits: string, letter: string): boolean {
  return LETTERS[parseInt(digits, 10) % 23] === letter
}

// reports whether a document is valid for the type it claims to be
export function isValidDocument(value: string, type: DocumentType): boolean {
  const v = normaliseDocument(value)
  if (type === 'dni') return DNI_RE.test(v) && validControlLetter(v.slice(0, 8), v[8]!)
  if (type === 'nie') {
    if (!NIE_RE.test(v)) return false
    return validControlLetter(NIE_PREFIX[v[0]!]! + v.slice(1, 8), v[8]!)
  }
  return PASSPORT_RE.test(v)
}
